contents/blog/clickhouse-materialized-columns.md

```diff
@@ -0,0 +1,134 @@
+
+From this we can see that the ClickHouse server CPU is spending most of its time parsing JSON.
+
+The typical solution would be to extract $current_url to a separate column. This would get rid of the JSON parsing and reduce the amount of data read from disk.
+
+However, in this particular case it wouldn’t work because:
+
+1. The data is passed from users - meaning we’d end up with millions (!) of unique columns
+2. This would complicate live data ingestion a lot, introducing new and exciting race conditions
+
+
+## Enter materialized columns
+
+Turns out, that's exactly the problems materialized columns can help solve.
+
+```sql
+ALTER TABLE events
+ADD COLUMN mat_$current_url
+VARCHAR MATERIALIZED JSONExtractString(properties_json, '$current_url')
+```
+
+This will create a new column that will be automatically filled for incoming data, creating a new file on disk. The data is automatically filled during `INSERT` statements, so data ingestion does not need to change.
+
+The trade-off is more data being stored on disk. In practice ClickHouse compresses data very well, making this a worthwhile trade-off. On our test dataset, mat_$current_url is only 1.5% the size of `properties_json` on disk with a 10x compression ratio. Other properties which have lower cardinality can achieve even better compression (we’ve seen up to 100x)!
+
+Just creating the column is not enough, since for old data queries would still resort to using a `JSONExtract`. For this reason, you want to backfill data. The easiest way currently is to run [OPTIMIZE](https://clickhouse.tech/docs/en/sql-reference/statements/optimize/) command:
+
+```sql
+OPTIMIZE TABLE events FINAL
+```
+
+After backfilling, running the updated query speeds things up significantly:
+
+```sql
+SELECT count(*)
+FROM events
+WHERE event = '$pageview'
+  AND mat_$current_url = 'https://app.posthog.com/login'
+  AND timestamp >= '2021-08-01'
+  AND timestamp < '2021-09-01'
+```
+
+Looking at `system.query_log`, the new query:
+
+- took 980ms (**71%/3.4x improvement**)
+- read 14.36 GiB from disk (**81%/5x improvement improvement**)
+
+The wins are even more magnified if more than one property filter is used at a time.
+
+
+## Backfilling efficiently
+
+Using `OPTIMIZE TABLE` after adding columns is often not a good idea, since it will involve a lot of I/O as the whole table gets rewritten.
+
+As of writing, there's a a feature request on [Github](https://github.com/ClickHouse/ClickHouse/issues/27730) for adding specific commands for this.
+
+As a work-around you can:
+1. Temporarily set the column to use `DEFAULT` instead of `MATERIALIZED`
+2. Backfill part of the data using `ALTER TABLE events UPDATE mat_$current_url = mat_$current_url WHERE timestamp >= '2021-08-01'`
+3. Set the column back to `MATERIALIZED`
+
+Note that hacky work-around has [implications](https://clickhouse.tech/docs/en/sql-reference/statements/create/table/#materialized) for your `INSERT` and `SELECT *` queries.
+
```

Comment:
There's a tradeoff between these two approaches. 

1. Using DEFAULT. Data will be materialized on a background merge.  Also, this approach might invalidate some SQL inserts because it is going to add a column that now appears as a normal table column. This would happen if you don't specify column names on the INSERT statement. 

2. Using MATERIALIZED. Materializes only on INSERT or OPTIMIZE TABLE as you pointed out, with the disadvantages of the latter.  You also can't do the ALTER TABLE UPDATE trick, again as you pointed out.

Reply:
Hi @hodgesrm, thanks for the feedback! 

I've updated the article. Good point regarding DEFAULT, didn't realize it would get materialized during merges in background. :)